### DISCUSSION TITLE
Docker linuxserver.io versioning

### DISCUSSION THREAD

**Discussion by Sebclem** [renovatebot/renovate#13052] [Request Help] [answered]
Hello,

Did anyone successfully configured renovate to upgrade linuxserver.io docker image ?

This is my current config:
```json
{
  "$schema": "https://docs.renovatebot.com/renovate-schema.json",
  "extends": [
    "config:base",
    "docker:enableMajor"
  ],
  "commitMessagePrefix": ":arrow_up:",
  "packageRules": [
    {
      "packagePatterns": ["^ghcr.io\\/linuxserver\\/"],
      "versionScheme": "regex:^(?<compatibility>.*(\\d+\\.)?)(?<major>\\d+)\\.(?<minor>\\d+)\\.(?<patch>\\d+)?$"
    }
  ]
}
```

It seems to correctly extract my version but it does not open PR to update my images (should upgrade `version-0.6.1341` to `version-0.6.1342`:
```json
{
   "depName": "ghcr.io/linuxserver/sonarr",
   "currentValue": "version-0.6.1341",
   "replaceString": "ghcr.io/linuxserver/sonarr:version-0.6.1341",
   "autoReplaceStringTemplate": "{{depName}}{{#if newValue}}:{{newValue}}{{/if}}{{#if newDigest}}@{{newDigest}}{{/if}}",
   "datasource": "docker",
   "depIndex": 0,
   "updates": [],
   "warnings": [],
   "versioning": "regex:^(?<compatibility>.*(\\d+\\.)?)(?<major>\\d+)\\.(?<minor>\\d+)\\.(?<patch>\\d+)?$",
   "sourceUrl": "https://github.com/linuxserver/docker-sonarr",
   "currentVersion": "version-0.6.1341",
   "fixedVersion": "version-0.6.1341"
 }
```

If you can find where the issue is or tell me if i'm doing it wrong :smiley: 

Thank !

**Comment by rarkins** [comment, marked as answer]
Your regex doesn't look right.

Try `regex:^(?<compatibility>.*?)-(?<major>\\d+)\\.(?<minor>\\d+)\\.(?<patch>\\d+)$`

You can test in regex101.com in future - just remember to replace single backslashes with doubles once you put it into your config.

**Reply by Sebclem** [reply]
Thanks ! It seams to work now !

**Reply by akhan23wgu** [reply]
I had to modify regex to the following in order to get RenovateBot to create a PR:

`"versioning": "regex:^(?<compatibility>.*?)-(?<major>\\d+)\\.(?<minor>\\d+)\\.(?<patch>\\d+)\\.(?<build>\\d+)$"`

**Reply by cliffordwhansen** [reply]
I found this one catches more of them:
```
"versionScheme": "regex:^(?<compatibility>.*?)-(?<major>v?\\d+)\\.(?<minor>\\d+)\\.(?<patch>\\d+)[\\.-]*(?<build>r?\\d+)*-*(?<tag>r?\w+)*$"
```

*edit to add tag at the end

**Reply by henyxia** [reply]
@cliffordwhansen , your edited regex lacks of on `\` on the tag.
Here is the fixed one:
```
"versionScheme": "regex:^(?<compatibility>.*?)-(?<major>v?\\d+)\\.(?<minor>\\d+)\\.(?<patch>\\d+)[\\.-]*(?<build>r?\\d+)*-*(?<tag>r?\\w+)*"
```

**Reply by viceice** [reply]
wrong too:
- the `r` can't be part of `build` group.
- `tag` group is not supported, use `release` and exclude `r` too

**Reply by cliffordwhansen** [reply]
@viceice thanks for the info
@henyxia I see I missed that in the paste (was correct in my config :face_with_spiral_eyes:)

```yaml
"versioning": "regex:^(?<compatibility>.*?)-(?<major>v?\\d+)\\.(?<minor>\\d+)\\.(?<patch>\\d+)[\\.-]*r?(?<build>\\d+)*-*r?(?<release>\\w+)*"
```

After updating to this version all my containers are now detected :smile: 

*edit* changed to "versioning"
*edit* updated to r? from r*

**Reply by viceice** [reply]
you should use `r?` otherwise it allows `rrrr` 😉

**Comment by mbrodala** [comment]
Notice that at least https://github.com/linuxserver/docker-openssh-server uses a completely different scheme which we now cover with this:

```json
  "packageRules": [
    {
      "matchPackageNames": ["ghcr.io/linuxserver/openssh-server"],
      "versioning": "regex:^(?<major>\\d+)\\.(?<minor>\\d+)_p(?<patch>\\d+)-(?<build>r\\d)-(?<revision>.+)$"
    }
  ]
```

**Reply by mbrodala** [reply]
That being said, for some reason Renovate does not suggest updates.

Currently we are on `9.7_p1-r4-ls163@sha256:0bdeba2d23dbe6772719c4510a5ec08c486dd315edcc24c68f09ca5c03013bce` for example and there are [a lot of updates](https://github.com/linuxserver/docker-openssh-server/pkgs/container/openssh-server), the latest being `9.7_p1-r4-ls178`.

How come Renovate does not suggest an update for the `revision`? In this case `ls163` to `ls178`.

I double-checked the RegEx once more and it seems fine:

![image](https://github.com/user-attachments/assets/59492c3f-78cd-4bc0-aad1-f661c5a18a70)

**Reply by mbrodala** [reply]
And right at this moment I got the error with a hint from @viceice above:

> the r can't be part of build group.

In other words: capture groups must result in purely numerical values. Only then Renovate can do a version comparison. String prefixes like `r` and `ls` are not supported, even if identical in all versions.

With this scheme it works:

```json
  "packageRules": [
    {
      "matchPackageNames": ["ghcr.io/linuxserver/openssh-server"],
      "versioning": "regex:^(?<major>\\d+)\\.(?<minor>\\d+)_p(?<patch>\\d+)-r(?<build>\\d)-ls(?<revision>.+)$"
    }
  ]
```

![image](https://github.com/user-attachments/assets/45be4712-0586-4d52-bfc5-081758aa01fe)

**Comment by teekennedy** [comment]
Looks like the versioning scheme has changed _again_ since this discussion was last updated. Here's the latest config that works for me:

```json
  "packageRules": [
    {
      "packagePatterns": ["^(gh|ls)cr\\.io\\/linuxserver\\/"],
      "versioning": "regex:^(?<major>\\d+)\\.(?<minor>\\d+)\\.(?<patch>\\d+)\\.(?<build>\\d+)-ls(?<revision>\\d+)$"
    }
  ]
```

Given that linuxserver publishes to both ghcr.io and their own lscr.io, I've also updated the `matchPackageNames` regex above to handle both.

**Comment by sidey79** [comment]
I have used the regex from @mbrodala  and @teekennedy  for the linuxserver/openssh-server images:


```
{
  "$schema": "https://docs.renovatebot.com/renovate-schema.json",
  "extends": [
    "config:recommended",
    "docker:pinDigests"
  ],
  "packageRules": [
    {
      "description": "Linuxserver IO openssh",
      "matchPackageNames": [
        "^(gh|ls)cr\\.io\\/linuxserver\\/openssh-server"
      ],
      "versioning": "regex:^(?<major>\\d+)\\.(?<minor>\\d+)_p(?<patch>\\d+)-r(?<build>\\d)-ls(?<revision>\\d+)"
    }
  ]
}
```

Renovate is not providing me pull requests
I checked the regex via regex101 and it should be the correct:
https://regex101.com/r/i7LouD/1

```
DEBUG: packageFiles with updates (repository=sidey79/docker-sftp-config, baseBranch=main)
       "config": {
         "docker-compose": [
           {
             "deps": [
               {
                 "depName": "ghcr.io/linuxserver/openssh-server",
                 "packageName": "ghcr.io/linuxserver/openssh-server",
                 "currentValue": "9.7_p1-r4-ls171",
                 "currentDigest": "sha256:53ea39d2485c0c96dc9abaae18cad4f29d0fcf33ad1c05f6d73b7f1e22ec215c",
                 "replaceString": "ghcr.io/linuxserver/openssh-server:9.7_p1-r4-ls171@sha256:53ea39d2485c0c96dc9abaae18cad4f29d0fcf33ad1c05f6d73b7f1e22ec215c",
                 "autoReplaceStringTemplate": "{{depName}}{{#if newValue}}:{{newValue}}{{/if}}{{#if newDigest}}@{{newDigest}}{{/if}}",
                 "datasource": "docker",
                 "updates": [],
                 "versioning": "docker",
                 "warnings": [],
                 "sourceUrl": "https://github.com/linuxserver/docker-openssh-server",
                 "registryUrl": "https://ghcr.io/",
                 "homepage": "https://github.com/linuxserver/docker-openssh-server/packages",
                 "lookupName": "linuxserver/openssh-server",
                 "currentVersion": "9.7_p1",
                 "fixedVersion": "9.7_p1-r4-ls171"
               }
             ],
             "packageFile": "docker-compose.yml"
           }
         ]
       }
```

Has anyone a idea, what i am missing here?

**Reply by teekennedy** [reply]
Your regex looks good to me, and good eye on fixing the literal dot in `.io` that I had in my config.

The only thing I can think of is that there's some kind of conflict with using the regex matcher with `docker:pinDigests`. I couldn't find anything relevant in the renovate docs though.

**Comment by sidey79** [comment]
I found the error in my config. Now the openssh server is detected.
The poblem was the matchPackageName, which wasn't really a compatible regex for renovate.

Thats a working packageRule for ghcr and lscr. 

```
  "packageRules": [
    {
      "description": "Linuxserver IO openssh",
      "matchPackageNames": ["/^(gh|ls)cr\\.io\\/linuxserver\\/openssh-server/"],
      "versioning": "regex:^(?<major>\\d+)\\.(?<minor>\\d+)_p(?<patch>\\d+)-r(?<build>\\d)-ls(?<revision>\\d+)"
    }
  ]
```

**Comment by jtcleek** [comment]
This works for me to match speedtest-tracker

```
  "packageRules": [
    {
      "matchPackageNames": ["lscr.io/linuxserver/speedtest-tracker"],
      "versioning": "regex:^v(?<major>\\d+)\\.(?<minor>\\d+)\\.(?<patch>\\d+)-ls(?<revision>\\d+)$"
    }
  ]
```

**Comment by davidwinter** [comment]
I've added a few extra variations on the versioning schemes I've found with packages. Currently using:

```json
"packageRules": [
    {
      "versioning": "regex:^v?(?<major>\\d+)\\.(?<minor>\\d+)\\.(?<patch>\\d+)((\\.|-r)(?<build>\\d+))?-ls(?<revision>\\d+)$",
      "matchPackageNames": [
        "/^(gh|ls)cr\\.io\\/linuxserver\\//"
      ]
    }
  ]
```

**Reply by yahoo1999** [reply]
Any idea how to deal with calibre-web? It apparently has some ancient builds tagged v5. Latest published version is [0.6.25-ls361](https://github.com/linuxserver/docker-calibre-web/releases/tag/0.6.25-ls361/) but it looks like renovate even with those package rules wants to update it to those v5 tags:
<img width="1202" height="199" alt="image" src="https://github.com/user-attachments/assets/97ec71b6-22ea-429c-905a-fa7f4628737b" />


EDIT: I tried to exclude v5 from allowed versions and so far it seems to be working:
```json
    {
      "allowedVersions": "!/^v5.*?$/",
      "matchPackageNames": ["lscr.io/linuxserver/calibre-web"]
    }
```
